src/main/kotlin/com/autonomousapps/internal/parse/KotlinBuildScriptDependenciesRewriter.kt

```diff
@@ -0,0 +1,176 @@
+
+/**
+ * Rewrites a Kotlin build script based on provided advice, which includes:
+ *
+ * - When entering the `dependencies` block, processing each dependency declaration to either remove or update it according to the advice provided.
+ * - Upon exiting the `dependencies` block, adding any new dependencies.
+ * - At the end of the script, adding a `dependencies` block if new dependencies are present and the block is currently missing.
+ *
+ * Note: Buildscript dependencies are ignored.
+ */
+internal class KotlinBuildScriptDependenciesRewriter(
+  private val input: CharStream,
+  private val tokens: CommonTokenStream,
+  private val errorListener: CollectingErrorListener,
+  private val advice: Set<Advice>,
+  private val printer: AdvicePrinter,
+  /** Reverse map from custom representation to standard. */
+  private val reversedDependencyMap: (String) -> String,
+) : BuildScriptDependenciesRewriter, KotlinParserBaseListener() {
+  private val rewriter = Rewriter(tokens)
+  private val indent = Whitespace.computeIndent(tokens, input)
+
+  private val dependencyExtractor = DependencyExtractor(
+    input = input,
+    tokens = tokens,
+    indent = indent,
+  )
+
+  private var hasDependenciesBlock = false
+  private var inBuildscriptBlock = false
+
+  /**
+   * Returns build script with advice applied.
+   *
+   * Throws [BuildScriptParseException] if the script has some idiosyncrasy that impairs parsing.
+   *
+   */
+  @Throws(BuildScriptParseException::class)
+  override fun rewritten(): String {
+    errorListener.getErrorMessages().ifNotEmpty {
+      throw BuildScriptParseException.withErrors(it)
+    }
+
+    return rewriter.text
+  }
+
+  override fun enterNamedBlock(ctx: NamedBlockContext) {
+    dependencyExtractor.onEnterBlock()
+
+    if (ctx.isBuildscript) {
+      inBuildscriptBlock = true
+    }
+
+    if (ctx.isDependencies) {
+      handleDependencies(ctx)
+    }
+  }
+
+  override fun exitNamedBlock(ctx: NamedBlockContext) {
+    dependencyExtractor.onExitBlock()
+
+    if (ctx.isDependencies && !inBuildscriptBlock) {
+      hasDependenciesBlock = true
+      advice.filterToSet { it.isAnyAdd() }.ifNotEmpty { addAdvice ->
+        val closeBrace = ctx.stop
+        rewriter.insertBefore(closeBrace, addAdvice.joinToString(separator = "\n", postfix = "\n") { a ->
+          printer.toDeclaration(a)
+        })
+      }
+    }
+
+    // Must be last
+    if (ctx.isBuildscript) {
+      inBuildscriptBlock = false
+    }
+  }
+
+  override fun exitScript(ctx: ScriptContext) {
+    // Exit early if this build script has a dependencies block. If it doesn't, we may need to add missing dependencies.
+    if (hasDependenciesBlock) return
+
+    advice.filterToOrderedSet { it.isAnyAdd() }.ifNotEmpty { addAdvice ->
+      rewriter.insertBefore(
+        ctx.stop,
+        addAdvice.joinToString(prefix = "\ndependencies {\n", postfix = "\n}\n", separator = "\n") { a ->
+          printer.toDeclaration(a)
+        }
+      )
+    }
```

Comment:
This block looks basically identical to the block above, with the exception that this one uses `filterToOrderedSet` and the above one is just `filterToSet`. I think We should extract this to a function and use the ordered set function.

Reply:
Done in https://github.com/autonomousapps/dependency-analysis-gradle-plugin/pull/1299/commits